File: solutions/scala/108-kth-largest-element-in-an-array.scala
/**
 * Problem 108: Kth Largest Element in an Array (LeetCode 215)
 * Difficulty: Med
 * Language: Scala
 */
def findKthLargest(nums: Array[Int], k: Int): Int = {
  val pivot = nums(nums.length / 2)
  val (hi, eq, lo) = (nums.filter(_ > pivot),
    nums.filter(_ == pivot), nums.filter(_ < pivot))
  if (k <= hi.length) findKthLargest(hi, k)
  else if (k <= hi.length + eq.length) pivot
  else findKthLargest(lo, k - hi.length - eq.length)
}

File: solutions/ocaml/108-kth-largest-element-in-an-array.ml
(*
  Problem 108: Kth Largest Element in an Array (LeetCode 215)
  Difficulty: Med
  Language: OCaml
*)
let rec find_kth_largest nums k =
  let pivot = List.nth nums (List.length nums / 2) in
  let hi = List.filter (fun x -> x > pivot) nums in
  let eq = List.filter (fun x -> x = pivot) nums in
  let lo = List.filter (fun x -> x < pivot) nums in
  if k <= List.length hi then find_kth_largest hi k
  else if k <= List.length hi + List.length eq then pivot
  else find_kth_largest lo (k - List.length hi - List.length eq)
